### PR TITLE
CURA-7090/Added check to reevaluate dismissed incompatible packages

### PR DESCRIPTION
**Pull request by DonExo**
EDIT: It was requested that this PR and the logic about it is somewhat reworked.
When the Next/Dismiss button in the right bottom corner is clicked - it automatically dismisses all incompatible packages, should there be any.
It puts them in a "dismissed" list in `packages.json`.

It also re-evaluates that list on every Cura startup (or Login)  and removes any package from the list in case it gets updated with the correct SDK version.

Needs: https://github.com/Ultimaker/Uranium/pull/568

### CODE REVIEW

File: plugins/Toolbox/src/CloudSync/SubscribedPackagesModel.py [hunk 1]
```diff
@@ -37,27 +37,22 @@ def hasIncompatiblePackages(self) -> bool:
...
             self.setProperty(package, property="is_dismissed", value=True)
             Logger.debug("Package {} has been dismissed".format(package_id))
 
-    def setMetadata(self, data: List[Dict[str, List[Any]]]) -> None:
-        self._metadata = data
-
     def addDiscrepancies(self, discrepancy: List[str]) -> None:
         self._discrepancies = discrepancy
 
     def getCompatiblePackages(self):
```
Comment (konskarm) [inline plugins/Toolbox/src/CloudSync/SubscribedPackagesModel.py]: Add return type

File: plugins/Toolbox/src/CloudSync/SubscribedPackagesModel.py [hunk 2]
```diff
@@ -37,27 +37,22 @@ def hasIncompatiblePackages(self) -> bool:
                 return True
         return False
 
-        # Sets the "is_compatible" to True for the given package, in memory
```
Comment (DonExo) [inline plugins/Toolbox/src/CloudSync/SubscribedPackagesModel.py]: This is probably a forgotten comment line which doesn't apply to the function beneath it.

File: plugins/Toolbox/src/CloudSync/DiscrepanciesPresenter.py [hunk 3]
```diff
@@ -28,13 +28,10 @@ def present(self, plugin_path: str, model: SubscribedPackagesModel) -> None:
...
-        self._package_manager.dismissPackage(package_id)  # adds this package_id as dismissed in the user config file
-
     def _onConfirmClicked(self, model: SubscribedPackagesModel) -> None:
         # For now, all compatible packages presented to the user should be installed.
         # Later, we might remove items for which the user unselected the package
+        if model.getIncompatiblePackages():
+            self._package_manager.dismissAllIncompatiblePackages(model.getIncompatiblePackages())
         model.setItems(model.getCompatiblePackages())
-        self.packageMutations.emit(model)
+        self.packageMutations.emit(model) # #### proveri sho e ova???
```
Comment (DonExo) [inline plugins/Toolbox/src/CloudSync/DiscrepanciesPresenter.py]:
```suggestion
        self.packageMutations.emit(model)
```

File: plugins/Toolbox/resources/qml/dialogs/CompatibilityDialog.qml [hunk 4]
```diff
@@ -20,6 +20,8 @@ UM.Dialog{
     maximumHeight: minimumHeight
     margin: 0
 
+    property string buttonText: subscribedPackagesModel.hasIncompatiblePackages && !subscribedPackagesModel.hasCompatiblePackages ? "Dismiss" : "Next"
```
[the remark this answers is not in the document]
Comment (DonExo) [reply plugins/Toolbox/resources/qml/dialogs/CompatibilityDialog.qml]: If you scroll a bit to the bottom, you can see that I'm using the catalog to translate the strings there.

`text: catalog.i18nc("@button", buttonText)`

Or are you saying this is not going to translate the, ?

File: plugins/Toolbox/src/CloudSync/CloudPackageChecker.py [hunk 5]
```diff
@@ -53,16 +78,12 @@ def _handleCompatibilityData(self, subscribed_packages_payload: List[Dict[str, A
...
-        package_discrepancy = list(set(user_subscribed_packages).difference(user_installed_packages))
-
-        self._model.addDiscrepancies(package_discrepancy) # TODO: Move these two lines below, under if package_discrepancy:
-        self._model.initialize(subscribed_packages_payload)
-
-        if not self._model.hasCompatiblePackages: # TODO: Remove these two lines...
-            return None
 
+        # We check if there are packages installed in Web Marketplace but not in Cura marketplace
+        package_discrepancy = list(set(user_subscribed_packages).difference(user_installed_packages))
```
[the remark this answers is not in the document]
Comment (DonExo) [reply plugins/Toolbox/src/CloudSync/CloudPackageChecker.py]: Thanks!

File: plugins/Toolbox/resources/qml/dialogs/CompatibilityDialog.qml (hunk 4 above)
[the remark this answers is not in the document]
Comment (fieldOfView) [reply plugins/Toolbox/resources/qml/dialogs/CompatibilityDialog.qml]: Don't do that! Only ever have string literals in the i18n*() calls.

The scripts used to extract the string literals to be translated do not work like that. The original strings will not be detected, so they will not be included in the strings that need to be translated.

[the remark this answers is not in the document]
Comment (DonExo) [reply plugins/Toolbox/resources/qml/dialogs/CompatibilityDialog.qml]: Thanks for the heads-up Aldo!

**Pull request merged**